webapp/channels/src/components/autosize_textarea.tsx

```diff
@@ -13,203 +14,169 @@ type Props = {
-        const textarea = this.textarea;
+        const scrollHeight = referenceRef.current.scrollHeight;
+        const currentTextarea = textarea.current;
 
-        if (height > 0 && height !== this.height) {
-            const style = getComputedStyle(textarea);
+        if (scrollHeight > 0 && scrollHeight !== height.current) {
+            const style = getComputedStyle(currentTextarea);
 
             // Directly change the height to avoid circular rerenders
-            textarea.style.height = `${height}px`;
+            currentTextarea.style.height = `${scrollHeight}px`;
 
-            this.height = height;
+            height.current = scrollHeight;
 
-            this.props.onHeightChange?.(height, parseInt(style.maxHeight || '0', 10));
+            onHeightChange?.(scrollHeight, parseInt(style.maxHeight || '0', 10));
         }
-    };
+    }
 
-    private recalculatePadding = () => {
-        if (!this.referenceRef.current || !this.textarea) {
+    const recalculateWidth = () => {
+        if (!measuringRef.current) {
             return;
         }
 
-        const textarea = this.textarea;
-        const {paddingRight} = getComputedStyle(textarea);
-
-        if (paddingRight && paddingRight !== this.referenceRef.current.style.paddingRight) {
-            this.referenceRef.current.style.paddingRight = paddingRight;
+        const width = measuringRef.current?.offsetWidth || -1;
+        if (width >= 0) {
+            window.requestAnimationFrame(() => {
+                onWidthChange?.(width);
+            });
         }
-    };
+    }
 
-    private recalculateWidth = () => {
-        if (!this.measuringRef) {
+    const recalculatePadding = () => {
+        if (!referenceRef.current || !textarea.current) {
             return;
         }
 
-        const width = this.measuringRef.current?.offsetWidth || -1;
-        if (width >= 0) {
-            window.requestAnimationFrame(() => {
-                this.props.onWidthChange?.(width);
-            });
+        const currentTextarea = textarea.current;
+        const {paddingRight} = getComputedStyle(currentTextarea);
+
+        if (paddingRight && paddingRight !== referenceRef.current.style.paddingRight) {
+            referenceRef.current.style.paddingRight = paddingRight;
         }
-    };
+    }
 
-    private setTextareaRef = (textarea: HTMLTextAreaElement) => {
-        if (this.props.forwardedRef) {
-            if (typeof this.props.forwardedRef === 'function') {
-                this.props.forwardedRef(textarea);
+    const setTextareaRef = useCallback((textareaRef: HTMLTextAreaElement) => {
+        if (ref) {
+            if (typeof ref === 'function') {
+                ref(textareaRef);
             } else {
-                this.props.forwardedRef.current = textarea;
+                ref.current = textareaRef;
             }
         }
 
-        this.textarea = textarea;
+        textarea.current = textareaRef;
+    }, [ref]);
+
+    useEffect(() => {
```

Comment:
Effects after callbacks and before rendering.

Reply:
Note that most `useEffect`s will have a dependency array so that they're not run on every render. This component is a special case because we want to recalculate the size more often